File: java/src/main/java/com/omnisync/core/resilience/CircuitBreaker.java
package com.omnisync.core.resilience;

import com.omnisync.core.error.CircuitBreakerOpenException;
import com.omnisync.core.error.NetworkException;
import com.omnisync.core.error.RateLimitExceededException;

import java.time.Duration;
import java.time.Instant;
import java.util.Objects;

/**
 * State machine protecting upstream services from cascading failures by short-circuiting requests.
 */
public class CircuitBreaker {

    public enum State {
        CLOSED,
        OPEN,
        HALF_OPEN
    }

    public static final int DEFAULT_FAILURE_THRESHOLD = 5;
    public static final Duration DEFAULT_RECOVERY_TIMEOUT = Duration.ofSeconds(30);
    public static final int DEFAULT_SUCCESS_THRESHOLD = 2;

    private final int failureThreshold;
    private final Duration recoveryTimeout;
    private final int successThreshold;

    private State state = State.CLOSED;
    private int failureCount = 0;
    private int halfOpenSuccessCount = 0;
    private Instant lastFailureTime;

    public CircuitBreaker() {
        this(DEFAULT_FAILURE_THRESHOLD, DEFAULT_RECOVERY_TIMEOUT, DEFAULT_SUCCESS_THRESHOLD);
    }

    public CircuitBreaker(int failureThreshold, Duration recoveryTimeout, int successThreshold) {
        this.failureThreshold = Math.max(1, failureThreshold);
        this.recoveryTimeout = Objects.requireNonNull(recoveryTimeout, "recoveryTimeout must not be null");
        this.successThreshold = Math.max(1, successThreshold);
    }

    /**
     * Evaluates circuit state, allowing execution or throwing CircuitBreakerOpenException.
     *
     * @throws CircuitBreakerOpenException if circuit is in OPEN state
     */
    public synchronized void allowExecution() throws CircuitBreakerOpenException {
        Instant now = Instant.now();

        if (state == State.OPEN) {
            Duration elapsed = Duration.between(lastFailureTime, now);
            if (elapsed.compareTo(recoveryTimeout) >= 0) {
                state = State.HALF_OPEN;
                halfOpenSuccessCount = 0;
            } else {
                Duration remaining = recoveryTimeout.minus(elapsed);
                throw new CircuitBreakerOpenException(
                        "Circuit breaker is OPEN. Requests short-circuited.",
                        remaining
                );
            }
        }
    }

    /**
     * Records a successful operation, potentially closing the circuit if in HALF_OPEN.
     */
    public synchronized void recordSuccess() {
        if (state == State.HALF_OPEN) {
            halfOpenSuccessCount++;
            if (halfOpenSuccessCount >= successThreshold) {
                state = State.CLOSED;
                failureCount = 0;
                halfOpenSuccessCount = 0;
            }
        } else if (state == State.CLOSED) {
            failureCount = 0;
        }
    }

    /**
     * Records an execution failure, opening the circuit if thresholds are exceeded.
     *
     * @param error the caught exception
     */
    public synchronized void recordFailure(Throwable error) {
        if (!(error instanceof NetworkException || error instanceof RateLimitExceededException)) {
            return;
        }

        Instant now = Instant.now();
        lastFailureTime = now;

        if (state == State.HALF_OPEN) {
            state = State.OPEN;
            failureCount = failureThreshold;
            halfOpenSuccessCount = 0;
        } else if (state == State.CLOSED) {
            failureCount++;
            if (failureCount >= failureThreshold) {
                state = State.OPEN;
            }
        }
    }

    public synchronized void reset() {
        state = State.CLOSED;
        failureCount = 0;
        halfOpenSuccessCount = 0;
        lastFailureTime = null;
    }

    public synchronized State getState() {
        return state;
    }

    public synchronized int getFailureCount() {
        return failureCount;
    }

    public Duration getRecoveryTimeout() {
        return recoveryTimeout;
    }
}

File: python/src/omnisync/resilience/circuit_breaker.py
"""Circuit breaker state machine preventing cascading system failures."""

from enum import Enum
import time
from typing import Optional

from omnisync.error.exceptions import (
    CircuitBreakerOpenError,
    NetworkError,
    RateLimitExceededError,
)


class CircuitState(str, Enum):
    """Possible operating states for the circuit breaker."""

    CLOSED = "CLOSED"
    OPEN = "OPEN"
    HALF_OPEN = "HALF_OPEN"


class CircuitBreaker:
    """State machine governing requests based on upstream failure rates."""

    def __init__(
        self,
        failure_threshold: int = 5,
        recovery_timeout_seconds: float = 30.0,
        success_threshold: int = 2,
    ) -> None:
        """Initialize CircuitBreaker.

        Args:
            failure_threshold: Consecutive failures to trip circuit to OPEN.
            recovery_timeout_seconds: Duration circuit stays OPEN before probe is allowed.
            success_threshold: Consecutive successes in HALF_OPEN to return to CLOSED.
        """
        self.failure_threshold = max(1, failure_threshold)
        self.recovery_timeout_seconds = max(0.0, recovery_timeout_seconds)
        self.success_threshold = max(1, success_threshold)

        self.state = CircuitState.CLOSED
        self.failure_count = 0
        self.half_open_success_count = 0
        self.last_failure_time: Optional[float] = None

    def allow_execution(self) -> None:
        """Verify circuit allows execution or raise CircuitBreakerOpenError.

        Raises:
            CircuitBreakerOpenError: When circuit is in OPEN state and recovery window has not elapsed.
        """
        now = time.monotonic()

        if self.state == CircuitState.OPEN:
            elapsed = now - (self.last_failure_time or 0.0)
            if elapsed >= self.recovery_timeout_seconds:
                self.state = CircuitState.HALF_OPEN
                self.half_open_success_count = 0
            else:
                remaining = self.recovery_timeout_seconds - elapsed
                raise CircuitBreakerOpenError(
                    "Circuit breaker is OPEN. Requests short-circuited.",
                    remaining_timeout_seconds=remaining,
                )

    def record_success(self) -> None:
        """Record successful execution."""
        if self.state == CircuitState.HALF_OPEN:
            self.half_open_success_count += 1
            if self.half_open_success_count >= self.success_threshold:
                self.state = CircuitState.CLOSED
                self.failure_count = 0
                self.half_open_success_count = 0
        elif self.state == CircuitState.CLOSED:
            self.failure_count = 0

    def record_failure(self, error: Exception) -> None:
        """Record an execution failure, updating state if threshold exceeded."""
        if not isinstance(error, (NetworkError, RateLimitExceededError)):
            return

        self.last_failure_time = time.monotonic()

        if self.state == CircuitState.HALF_OPEN:
            self.state = CircuitState.OPEN
            self.failure_count = self.failure_threshold
            self.half_open_success_count = 0
        elif self.state == CircuitState.CLOSED:
            self.failure_count += 1
            if self.failure_count >= self.failure_threshold:
                self.state = CircuitState.OPEN

    def reset(self) -> None:
        """Reset circuit breaker to initial CLOSED state."""
        self.state = CircuitState.CLOSED
        self.failure_count = 0
        self.half_open_success_count = 0
        self.last_failure_time = None
